File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/FavoriteIterator.cs
namespace IteratorDesignPattern.Example3;

public class FavoriteIterator : IIterator<Song>
{
    private readonly List<Song> _songs;
    private int _currentIndex = 0;

    public FavoriteIterator(FavoritePlaylist playlist)
    {
        _songs = playlist.GetFavorites().ToList();
    }

    public bool HasNext()
    {
        return _currentIndex < _songs.Count;
    }

    public Song Next()
    {
        return _songs[_currentIndex++];
    }

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/src/main/java/behavioral/iterator/example3/FavoriteIterator.java
package behavioral.iterator.example3;

import java.util.*;

public class FavoriteIterator implements IIterator<Song> {
    private final List<Song> songs;
    private int currentIndex = 0;

    public FavoriteIterator(FavoritePlaylist playlist) {
        songs = new ArrayList<>(playlist.getFavorites());
    }

    @Override
    public boolean hasNext() { return currentIndex < songs.size(); }

    @Override
    public Song next() { return songs.get(currentIndex++); }

    @Override
    public void reset() { currentIndex = 0; }
}
